lib/client/aws_role/impl.go

```diff
@@ -0,0 +1,135 @@
+package aws_role
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	"github.com/Cloud-Foundations/keymaster/lib/paths"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/aws/arn"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+)
+
+func parseArn(arnString string) (*arn.ARN, error) {
+	parsedArn, err := arn.Parse(arnString)
+	if err != nil {
+		return nil, err
+	}
+	switch parsedArn.Service {
+	case "iam", "sts":
+	default:
+		return nil, fmt.Errorf("unsupported service: %s", parsedArn.Service)
+	}
+	splitResource := strings.Split(parsedArn.Resource, "/")
+	if len(splitResource) < 2 || splitResource[0] != "assumed-role" {
+		return nil, fmt.Errorf("invalid resource: %s", parsedArn.Resource)
+	}
+	parsedArn.Service = "iam"
+	parsedArn.Resource = "role/" + splitResource[1]
+	return &parsedArn, nil
+}
+
+// Returns certificate PEM block.
+func (p Params) getRoleCertificate() ([]byte, error) {
+	if p.KeymasterServer == "" {
+		return nil, fmt.Errorf("no keymaster server specified")
+	}
+	if p.Logger == nil {
+		return nil, fmt.Errorf("no logger specified")
+	}
+	if p.Context == nil {
+		p.Context = context.TODO()
+	}
+	if p.HttpClient == nil {
+		p.HttpClient = http.DefaultClient
+	}
+	if p.Signer == nil {
+		signer, err := rsa.GenerateKey(rand.Reader, 2048)
+		if err != nil {
+			return nil, err
+		}
+		p.Signer = signer
+		p.KeyType = "RSA"
+	}
+	if p.KeyType != "RSA" {
+		return nil, fmt.Errorf("unsupported key type: %s", p.KeyType)
+	}
+	derPubKey, err := x509.MarshalPKIXPublicKey(p.Signer.Public())
```

Comment:
This was simple. The caller can provider their own signer if desired.
Note also the `rsaKeySize` constant from the keymaster `main` package can't be used since this is a separate package.